functorch/experimental/_nonzero_static.py

```diff
@@ -0,0 +1,65 @@
+import torch
+import torch.library
+
+
+"""
+Define a `nonzero_static` operation.
+"""
+
+
+def nonzero_static(self: torch.Tensor, *, size: int, fill_value=-1):
+    """
+    Returns a 2-D tensor where each row is the index for a nonzero value.
+
+    The number of rows will alway be `size`.
+    If the actual number of non-zero elements is smaller than `size`, the return
+    value will be padded with 'fill_value'.
+    If the actual number of non-zero elements is larger than `size`, the return
+    value will be truncated.
+
+    For example,
+
+    x = torch.tensor([[1, 0], [3, 2]])
+
+    call regular `nonzero()` on x will give you a tensor `tensor([[0, 0], [1, 0], [1, 1])`
+
+    call `nonzero_static(x, size=4)` on x will give you a tensor
+    `tensor([[0, 0], [1, 0], [1, 1], [fill_value, fill_value])` (padded)
+
+    call `nonzero_static(x, size=2)` on x will give you a tensor
+    `tensor([[0, 0], [1, 0])` (truncated)
+    """
+    # Check if 'fill_value' is an integer type
+    assert isinstance(fill_value, int), "'fill_value' must be an integer"
+    # Check if size is an non-negative integer
+    assert size >= 0, "'size' must be an non-negative integer"
+    # Check if self is not 0-d tensor
+    assert self.dim() > 0, "Rank must be positive"
+
+    if size == 0:
+        # Return empty tensor
+        return torch.empty(0, dtype=torch.long)
+
+    # Delegate call to regular nonzero to get a data-dependent output
+    dyn_result = self.nonzero(as_tuple=False)
+    # Allocate a fixed-size out tensor with padding
+    ndim = self.dim()
+    num_nonzero = dyn_result.size(0)
+    result = torch.full((size, ndim), fill_value, dtype=torch.long)
+    copy_size = min(size, num_nonzero)
+    result[:copy_size, :] = dyn_result[:copy_size, :]
+    return result
+
+
+experimental_ops = torch.library.Library("experimental", "DEF")
+experimental_ops.define(
+    "nonzero_static(Tensor self, *, int size, Scalar fill_value=-1) -> Tensor"
+)
+experimental_ops.impl("nonzero_static", nonzero_static, "CompositeExplicitAutograd")
+
+
+def meta_nonzero_static(self: torch.Tensor, *, size: int, fill_value=-1):
+    return self.new_empty((size, self.dim()))
```

Comment:
The result is always an int tensor, even if the input is a float tensor, so new_empty here is wrong